org-code-javabuilder/lib/src/main/java/org/code/javabuilder/ProjectFileLoader.java

```diff
@@ -0,0 +1,66 @@
+package org.code.javabuilder;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.concurrent.Callable;
+
+/**
+ * Fetches an individual file from a student's project. This should be used with the
+ * {dashboardHost}/v3/files/{channelId}/{fileName} API. It has no permissions to use AWS S3. This
+ * should be used asynchronously so multiple files can be retrieved at once.
+ */
+public class ProjectFileLoader implements Callable<Boolean> {
+  private final ProjectFile projectFile;
+  private final String fileUrl;
+
+  /**
+   * @param projectFile The in-memory representation of the given file
+   * @param fileUrl The URL to use to fetch the file
+   */
+  public ProjectFileLoader(ProjectFile projectFile, String fileUrl) {
+    this.projectFile = projectFile;
+    this.fileUrl = fileUrl;
+  }
+
+  /**
+   * @return true if the load succeeds
+   * @throws UserFacingException If any errors occur while fetching the file.
+   */
+  @Override
+  public Boolean call() throws UserFacingException {
+    try {
+      URL url = new URL(fileUrl);
```

Comment:
This seems really low-level, perhaps [java.net.http.HttpClient](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html) or Apache's [HttpClient](https://hc.apache.org/httpcomponents-client-5.0.x/index.html) would be a better choice?  I think both of these also support async call patterns so you might be able to get rid of this class entirely.

Reply:
Good call. Also, with the way it's looking like we'll change things for multi-file, we may only need to do one HTTP request too.